compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -32,25 +32,27 @@ use crate::{
         UnaryOp, UnresolvedType, UnresolvedTypeData,
     },
     node_interner::{DefinitionKind, ExprId, GlobalId, TraitId, TraitImplKind, TraitMethodId},
-    Generics, ResolvedGeneric, Type, TypeBinding, TypeKind, TypeVariable, TypeVariableKind,
+    Generics, Kind, ResolvedGeneric, Type, TypeBinding, TypeVariable, TypeVariableKind,
 };
 
 use super::{lints, Elaborator};
 
 impl<'context> Elaborator<'context> {
     /// Translates an UnresolvedType to a Type with a `TypeKind::Normal`
     pub(super) fn resolve_type(&mut self, typ: UnresolvedType) -> Type {
-        self.resolve_type_inner(typ, TypeKind::Normal)
-    }
-
-    /// Translate an UnresolvedType to a Type with a `TypeKind::Numeric`
-    pub(super) fn resolve_numeric_type(&mut self, typ: UnresolvedType) -> Type {
-        self.resolve_type_inner(typ, TypeKind::Numeric)
+        let span = typ.span;
+        let resolved_type = self.resolve_type_inner(typ, &Kind::Normal);
+        if resolved_type.is_nested_slice() {
+            self.push_err(ResolverError::NestedSlices {
+                span: span.expect("Type should have span"),
+            });
+        }
+        resolved_type
     }
 
     /// Translates an UnresolvedType into a Type and appends any
     /// freshly created TypeVariables created to new_variables.
-    pub fn resolve_type_inner(&mut self, typ: UnresolvedType, kind: TypeKind) -> Type {
+    pub fn resolve_type_inner(&mut self, typ: UnresolvedType, kind: &Kind) -> Type {
```

Comment:
We take in the kind here but we never actually check the returned type is of the expected kind

Reply:
I was only checking for the `NamedGeneric` case right now as it is the only type with a `Kind` at the moment. Where else were you thinking to check that the returned type is off the expected kind?

Reply:
At the bottom of `resolve_type_inner` before a type is returned. That way we can ensure a Type::Constant is not a type, or a function type is not a numeric generic.

Reply:
A global would resolve to a `Type::Constant` holding it's value, so we'll need to ensure a numeric type is expected in that case. The NamedGeneric check wouldn't catch this case since it is not a NamedGeneric. We'll need to check every type, not just generics.

Reply:
Discussed on slack, this check will be generalized in a follow-up when we remove implicit numeric generics https://github.com/noir-lang/noir/issues/5156